File: static/cover-letter-generator.js
console.log('Hire Ready cover-letter-generator.js loaded');

(function () {
  const API_BASE = 'https://resume-writer.onrender.com';

  function getToken() {
    return localStorage.getItem('hire_ready_token');
  }

  function escapeHtml(value) {
    if (value === null || value === undefined) return '';
    return String(value)
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&#039;');
  }

  async function hireReadyFetch(url, options = {}) {
    const token = getToken();
    let response = await fetch(url, {
      ...options,
      headers: {
        ...(options.headers || {}),
        Authorization: 'Bearer ' + token
      }
    });

    if (response.status !== 401) return response;

    const refreshToken = localStorage.getItem('hire_ready_refresh_token');
    if (!refreshToken) return response;

    const refreshResponse = await fetch(`${API_BASE}/api/auth/refresh`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ refresh_token: refreshToken })
    });

    if (!refreshResponse.ok) return response;

    const refreshData = await refreshResponse.json();
    localStorage.setItem('hire_ready_token', refreshData.access_token);
    localStorage.setItem('hire_ready_refresh_token', refreshData.refresh_token);
    localStorage.setItem('hire_ready_user', JSON.stringify(refreshData.user));
    localStorage.setItem('hire_ready_tier', refreshData.user.tier);

    return fetch(url, {
      ...options,
      headers: {
        ...(options.headers || {}),
        Authorization: 'Bearer ' + refreshData.access_token
      }
    });
  }

  function setStatus(message, type) {
    const status = document.getElementById('clg-status');
    if (!status) return;
    status.className = `clg-status ${type || ''}`;
    status.innerHTML = escapeHtml(message || '');
  }

  function setLoading(isLoading) {
    const button = document.getElementById('clg-submit');
    if (!button) return;
    button.disabled = isLoading;
    button.innerHTML = isLoading ? 'Generating...' : 'Generate Cover Letter';
  }

  function renderUpgrade(message) {
    const upgrade = document.getElementById('clg-upgrade');
    if (!upgrade) return;
    upgrade.innerHTML = `
      <div class="clg-upgrade">
        <strong>Premium feature</strong>
        <p>${escapeHtml(message || 'Upgrade to Premium to generate tailored cover letters from scratch.')}</p>
        <a href="/pricing/" class="clg-btn">View Pricing</a>
      </div>
    `;
  }

  function renderResults(data) {
    const results = document.getElementById('clg-results');
    if (!results) return;

    const analysis = data.analysis || {};
    results.classList.add('active');
    results.innerHTML = `
      <div class="clg-card">
        <h2>Your Generated Cover Letter</h2>

        <div class="clg-score-grid">
          <div class="clg-score">
            <span>${escapeHtml(analysis.overall_score || 0)}</span>
            <small>Overall</small>
          </div>
          <div class="clg-score">
            <span>${escapeHtml(analysis.ats_score || 0)}</span>
            <small>ATS</small>
          </div>
          <div class="clg-score">
            <span>${escapeHtml(analysis.job_alignment_score || 0)}</span>
            <small>Role Match</small>
          </div>
        </div>

        <pre class="clg-output">${escapeHtml(data.cover_letter || '')}</pre>
      </div>
    `;

    results.scrollIntoView({ behavior: 'smooth', block: 'start' });
  }

  async function checkCanRun() {
    if (!getToken()) {
      setStatus('Please log in to use the Cover Letter Generator.', 'error');
      return;
    }

    try {
      const response = await hireReadyFetch(`${API_BASE}/api/cover-letter-generator/can-run`);
      const data = await response.json();
      if (!response.ok || !data.success) return;
      if (!data.can_run) {
        setStatus(data.message || 'Upgrade required.', 'error');
        renderUpgrade(data.message);
      }
    } catch (error) {
      console.warn('Could not check generator usage', error);
    }
  }

  async function generateCoverLetter(event) {
    event.preventDefault();

    if (!getToken()) {
      setStatus('Please log in to use the Cover Letter Generator.', 'error');
      return;
    }

    const payload = {
      title: document.getElementById('clg-title')?.value || 'Generated Cover Letter',
      applicant_name: document.getElementById('clg-applicant-name')?.value || '',
      target_role: document.getElementById('clg-target-role')?.value || '',
      company_name: document.getElementById('clg-company')?.value || null,
      tone_preference: document.getElementById('clg-tone')?.value || 'professional',
      job_posting: document.getElementById('clg-job-posting')?.value || '',
      experience: document.getElementById('clg-experience')?.value || null,
      achievements: document.getElementById('clg-achievements')?.value || null
    };

    if (payload.applicant_name.trim().length < 2) {
      setStatus('Please enter the applicant name.', 'error');
      return;
    }

    if (payload.target_role.trim().length < 2) {
      setStatus('Please enter the target role.', 'error');
      return;
    }

    if (payload.job_posting.trim().length < 50) {
      setStatus('Please paste a fuller job advertisement or role description.', 'error');
      return;
    }

    setLoading(true);
    setStatus('Generating your tailored cover letter...', '');

    try {
      const response = await hireReadyFetch(`${API_BASE}/api/cover-letter-generator/generate`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const data = await response.json();

      if (!response.ok || !data.success) {
        setStatus(data.error || data.detail || 'Cover letter generation failed.', 'error');
        if (data.upgrade_required) renderUpgrade(data.message || data.error);
        return;
      }

      setStatus('Cover letter generated and saved successfully.', 'success');
      renderResults(data);
    } catch (error) {
      console.error('Cover letter generator error:', error);
      setStatus('Something went wrong. Please try again.', 'error');
    } finally {
      setLoading(false);
    }
  }

  document.addEventListener('DOMContentLoaded', function () {
    const form = document.getElementById('clg-form');
    if (form) {
      form.addEventListener('submit', generateCoverLetter);
      checkCanRun();
    }
  });
})();

File: routes/cover_letter_generator.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from fastapi.encoders import jsonable_encoder
from fastapi.responses import JSONResponse
from pydantic import BaseModel

from routes.user_management import get_current_user
from routes.cover_letter import ai_analyze_cover_letter
from routes.cover_letter_helpers import ai_generate_cover_letter, ai_retarget_cover_letter
from app.services.cover_letter_generator_service import (
    can_run_cover_letter_generator,
    get_cover_letter_generation,
    increment_cover_letter_generator_usage,
    list_cover_letter_generations,
    save_cover_letter_generation,
)

router = APIRouter()


class CoverLetterGeneratorRequest(BaseModel):
    title: Optional[str] = None
    applicant_name: str
    target_role: str
    company_name: Optional[str] = None
    job_posting: str
    experience: Optional[str] = None
    achievements: Optional[str] = None
    tone_preference: Optional[str] = "professional"


class CoverLetterRetargetRequest(BaseModel):
    title: Optional[str] = None
    applicant_name: Optional[str] = None
    source_cover_letter: str
    target_role: str
    company_name: Optional[str] = None
    job_posting: str
    tone_preference: Optional[str] = "professional"


@router.get("/cover-letter-generator/health")
async def cover_letter_generator_health():
    return {"status": "healthy", "service": "cover-letter-generator"}


@router.get("/cover-letter-generator/can-run")
async def can_run_generator(current_user: dict = Depends(get_current_user)):
    usage_status = can_run_cover_letter_generator(current_user)
    return {"success": True, **usage_status}


@router.post("/cover-letter-generator/generate")
async def generate_cover_letter(payload: CoverLetterGeneratorRequest, current_user: dict = Depends(get_current_user)):
    usage_status = can_run_cover_letter_generator(current_user)
    if not usage_status.get("can_run"):
        return JSONResponse(
            status_code=403,
            content=jsonable_encoder({
                "success": False,
                "error": usage_status.get("message"),
                **usage_status,
            }),
        )

    if len((payload.applicant_name or "").strip()) < 2:
        raise HTTPException(status_code=400, detail="Applicant name is required")
    if len((payload.target_role or "").strip()) < 2:
        raise HTTPException(status_code=400, detail="Target role is required")
    if len((payload.job_posting or "").strip()) < 50:
        raise HTTPException(status_code=400, detail="Please paste a fuller job advertisement or role description")

    try:
        generated_cover_letter = await ai_generate_cover_letter(
            job_posting=payload.job_posting,
            applicant_name=payload.applicant_name,
            current_role=payload.target_role,
            experience=payload.experience,
            achievements=payload.achievements,
            company_name=payload.company_name,
            tone_preference=payload.tone_preference or "professional",
        )

        analysis = await ai_analyze_cover_letter(
            cover_letter_text=generated_cover_letter,
            target_role=payload.target_role,
            job_posting=payload.job_posting,
            company_name=payload.company_name,
        )

        saved_result = save_cover_letter_generation(
            user_id=current_user["user_id"],
            title=payload.title,
            applicant_name=payload.applicant_name,
            target_role=payload.target_role,
            company_name=payload.company_name,
            job_posting=payload.job_posting,
            experience=payload.experience,
            achievements=payload.achievements,
            tone_preference=payload.tone_preference or "professional",
            generated_cover_letter=generated_cover_letter,
            analysis=analysis,
        )

        increment_cover_letter_generator_usage(current_user["user_id"])
        updated_usage = can_run_cover_letter_generator(current_user)

        return JSONResponse(content=jsonable_encoder({
            "success": True,
            "generation_id": saved_result.get("generation_id"),
            "cover_letter": generated_cover_letter,
            "analysis": analysis,
            "saved_result": saved_result,
            "usage": updated_usage,
        }))

    except HTTPException:
        raise
    except Exception as error:
        print(f"❌ Cover letter generator error: {str(error)}")
        return JSONResponse(
            status_code=500,
            content={
                "success": False,
                "error": f"Cover letter generation failed: {str(error)}",
            },
        )


@router.post("/cover-letter-generator/retarget")
async def retarget_cover_letter(payload: CoverLetterRetargetRequest, current_user: dict = Depends(get_current_user)):
    usage_status = can_run_cover_letter_generator(current_user)
    if not usage_status.get("can_run"):
        return JSONResponse(
            status_code=403,
            content=jsonable_encoder({
                "success": False,
                "error": usage_status.get("message"),
                **usage_status,
            }),
        )

    if len((payload.source_cover_letter or "").strip()) < 50:
        raise HTTPException(status_code=400, detail="Please provide a fuller existing cover letter to retarget")
    if len((payload.target_role or "").strip()) < 2:
        raise HTTPException(status_code=400, detail="New target role is required")
    if len((payload.job_posting or "").strip()) < 50:
        raise HTTPException(status_code=400, detail="Please paste a fuller job advertisement or role description")

    try:
        retargeted_cover_letter = await ai_retarget_cover_letter(
            source_cover_letter=payload.source_cover_letter,
            job_posting=payload.job_posting,
            target_role=payload.target_role,
            company_name=payload.company_name,
            tone_preference=payload.tone_preference or "professional",
        )

        analysis = await ai_analyze_cover_letter(
            cover_letter_text=retargeted_cover_letter,
            target_role=payload.target_role,
            job_posting=payload.job_posting,
            company_name=payload.company_name,
        )

        saved_result = save_cover_letter_generation(
            user_id=current_user["user_id"],
            title=payload.title or f"Retargeted Cover Letter - {payload.target_role}",
            applicant_name=payload.applicant_name,
            target_role=payload.target_role,
            company_name=payload.company_name,
            job_posting=payload.job_posting,
            experience="Retargeted from existing cover letter",
            achievements=None,
            tone_preference=payload.tone_preference or "professional",
            generated_cover_letter=retargeted_cover_letter,
            analysis=analysis,
        )

        increment_cover_letter_generator_usage(current_user["user_id"])
        updated_usage = can_run_cover_letter_generator(current_user)

        return JSONResponse(content=jsonable_encoder({
            "success": True,
            "mode": "retarget",
            "generation_id": saved_result.get("generation_id"),
            "cover_letter": retargeted_cover_letter,
            "source_cover_letter": payload.source_cover_letter,
            "analysis": analysis,
            "saved_result": saved_result,
            "usage": updated_usage,
        }))

    except HTTPException:
        raise
    except Exception as error:
        print(f"❌ Cover letter retarget error: {str(error)}")
        return JSONResponse(
            status_code=500,
            content={
                "success": False,
                "error": f"Cover letter retargeting failed: {str(error)}",
            },
        )


@router.get("/cover-letter-generator/history")
async def generator_history(current_user: dict = Depends(get_current_user)):
    return {
        "success": True,
        "results": list_cover_letter_generations(current_user["user_id"]),
    }


@router.get("/cover-letter-generator/{generation_id}")
async def get_generator_result(generation_id: str, current_user: dict = Depends(get_current_user)):
    result = get_cover_letter_generation(current_user["user_id"], generation_id)
    if not result:
        raise HTTPException(status_code=404, detail="Cover letter generation result not found")
    return {"success": True, "result": result}
